tests/phpunit/tests/l10n/wpWordCount.php

```diff
@@ -0,0 +1,101 @@
+		$settings = array(
+			'shortcodes' => array( 'shortcode' ),
+		);
+
+		$this->assertEquals( wp_word_count( $string, 'words', $settings ), $words );
+		$this->assertEquals( wp_word_count( $string, 'characters_excluding_spaces', $settings ), $characters_excluding_spaces );
+		$this->assertEquals( wp_word_count( $string, 'characters_including_spaces', $settings ), $characters_including_spaces );
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array
+	 */
+	public function data_get_string_variations() {
+		return array(
+			'Basic test'     => array(
+				'string'                      => 'one two three',
+				'words'                       => 3,
+				'characters_excluding_spaces' => 11,
+				'characters_including_spaces' => 13,
+			),
+			'HTML tags'      => array(
+				'string'                      => 'one <em class="test">two</em><br />three',
+				'words'                       => 3,
+				'characters_excluding_spaces' => 11,
+				'characters_including_spaces' => 12,
+			),
+			'Line breaks'    => array(
+				'string'                      => "one\ntwo\nthree",
+				'words'                       => 3,
+				'characters_excluding_spaces' => 11,
+				'characters_including_spaces' => 11,
+			),
+			'Encoded spaces' => array(
+				'string'                      => 'one&nbsp;two&#160;three',
+				'words'                       => 3,
+				'characters_excluding_spaces' => 11,
+				'characters_including_spaces' => 13,
+			),
+			'Punctuation'    => array(
+				'string'                      => "It's two three " . json_decode( '"\u2026"' ) . ' 4?',
+				'words'                       => 3,
+				'characters_excluding_spaces' => 15,
+				'characters_including_spaces' => 19,
+			),
+			'Em dash'        => array(
+				'string'                      => 'one' . json_decode( '"\u2014"' ) . 'two--three',
+				'words'                       => 3,
+				'characters_excluding_spaces' => 14,
+				'characters_including_spaces' => 14,
+			),
+			'Shortcodes'     => array(
+				'string'                      => 'one [shortcode attribute="value"]two[/shortcode]three',
+				'words'                       => 3,
+				'characters_excluding_spaces' => 11,
+				'characters_including_spaces' => 12,
+			),
+			'Astrals'        => array(
+				'string'                      => json_decode( '"\uD83D\uDCA9"' ),
+				'words'                       => 1,
+				'characters_excluding_spaces' => 1,
+				'characters_including_spaces' => 1,
+			),
+			'HTML comment'   => array(
+				'string'                      => 'one<!-- comment -->two three',
+				'words'                       => 2,
+				'characters_excluding_spaces' => 11,
+				'characters_including_spaces' => 12,
+			),
+			'HTML entity'    => array(
+				'string'                      => '&gt; test',
+				'words'                       => 1,
+				'characters_excluding_spaces' => 5,
+				'characters_including_spaces' => 6,
+			),
+		);
```

Comment:
This should also include datasets for:

- empty text: `''`
- text containing only whitespace: `"\t\r\n "`

There should be additional test methods added to the PR to cover:
- A `$type` that is not `words`, `characters_excluding_spaces` or `characters_including_spaces`, to ensure that this defaults to `words` per the expected behaviour.
- A non-array `$settings['shortcodes']`.
- An empty `$settings['shortcodes']`.
- A test in which the final `preg_match_all()` returns no matches.

Reply:
> * empty text: `''`
> * text containing only whitespace: `"\t\r\n "`

Added to the data provider.
 
> * A `$type` that is not `words`, `characters_excluding_spaces` or `characters_including_spaces`, to ensure that this defaults to `words` per the expected behaviour.

Added as `test_wp_word_count_should_use_the_default_word_count_type()` method.

> * A non-array `$settings['shortcodes']`.
> * An empty `$settings['shortcodes']`.

Added `test_wp_word_count_containing_non_array_shortcode_setting` and `test_wp_word_count_containing_empty_array_shortcode_setting` methods.

At the same time, I discovered that there is no test to confirm that the shortcode is remove, so I added it as `text containing a shortcode` to the data provider.

> * A test in which the final `preg_match_all()` returns no matches.

As far as I have tested, I have confirmed that `preg_match_all()` always returns an array, as shown below:

![preg_match_all](https://user-images.githubusercontent.com/54422211/236839758-2bbdd45d-d86d-4dea-a614-0b5f0c3441c9.png)

Should it be rewritten as follows?

```diff
diff --git a/src/wp-includes/l10n.php b/src/wp-includes/l10n.php
index 8a399ae638..ef0ff42b66 100644
--- a/src/wp-includes/l10n.php
+++ b/src/wp-includes/l10n.php
@@ -1937,9 +1937,5 @@ function wp_word_count( $text, $type, $settings = array() ) {
        // Match with the selected type regular expression to count the items.
        preg_match_all( $settings[ $type . '_regexp' ], $text, $matches );
 
-       if ( $matches ) {
-               return count( $matches[0] );
-       }
-
-       return $count;
+       return count( $matches[0] );
 }
```

Reply:
Fixed by 4fe0576